nucypher/policy/conditions/auth.py

```diff
@@ -0,0 +1,86 @@
+from enum import Enum
+from typing import List
+
+from eth_account.account import Account
+from eth_account.messages import HexBytes, encode_typed_data
+from siwe import SiweMessage, VerificationError
+
+
+class Auth:
+    class AuthScheme(Enum):
+        EIP712 = "EIP712"
+        SIWE = "SIWE"
+
+        @classmethod
+        def values(cls) -> List[str]:
+            return [scheme.value for scheme in cls]
+
+    class InvalidData(Exception):
+        pass
+
+    class AuthenticationFailed(Exception):
+        pass
+
+    @classmethod
+    def authenticate(cls, data, signature, expected_address):
+        raise NotImplementedError
+
+    @classmethod
+    def from_scheme(cls, scheme: str):
+        if scheme == cls.AuthScheme.EIP712.value:
+            return EIP712Auth
+        elif scheme == cls.AuthScheme.SIWE.value:
+            return SIWEAuth
+
+        raise ValueError(f"Invalid authentication scheme: {scheme}")
+
+
+class EIP712Auth(Auth):
+    @classmethod
+    def authenticate(cls, data, signature, expected_address):
+        try:
+            # convert hex data for byte fields - bytes are expected by underlying library
+            # 1. salt
+            salt = data["domain"]["salt"]
+            data["domain"]["salt"] = HexBytes(salt)
+            # 2. blockHash
+            blockHash = data["message"]["blockHash"]
+            data["message"]["blockHash"] = HexBytes(blockHash)
+
+            signable_message = encode_typed_data(full_message=data)
+            address_for_signature = Account.recover_message(
+                signable_message=signable_message, signature=signature
+            )
+        except Exception as e:
+            # data could not be processed
+            raise cls.InvalidData(f"Invalid auth data: {e.__class__.__name__} - {e}")
+
+        if address_for_signature != expected_address:
+            # verification failed - addresses don't match
+            raise cls.AuthenticationFailed(
+                f"Invalid EIP712 signature; does not match expected address, {expected_address}"
+            )
+
+
+class SIWEAuth(Auth):
+    @classmethod
+    def authenticate(cls, data, signature, expected_address):
```

Comment:
I see this method is similar to this `validate` method in SIWE example from login.xyz: https://github.com/spruceid/siwe/blob/main/packages/siwe/lib/client.ts#L195

A more complete example that actually checks the data in `typedData`, not just signature, can be seen here: https://github.com/spruceid/siwe/blob/main/packages/siwe/lib/client.ts#L211

Reply:
If I'm understanding what you mean, the answer depends on what is in the message - is the SIWE template message constant (eg. the domain is `taco`), or can it vary.  If it varies then would checking those things be more of add-on auth conditions that would need to be used - that's what I was referring to here - https://github.com/nucypher/nucypher/pull/3502#discussion_r1603258091. Is checking `domain`, `resources` etc. something that is done by default, or are add-on authentication conditions - eg. the following from lit - https://developer.litprotocol.com/v2/accesscontrol/evm/siwe#domain. 

That depends on what is done in the client - do you have thoughts there?

Reply:
> A more complete example that actually checks the data in typedData, not just signature, can be seen here:
> https://github.com/spruceid/siwe/blob/main/packages/siwe/lib/client.ts#L211

Just to clarify, `verify` is used in siwe-py (not the equivalent of `validate`)- https://github.com/spruceid/siwe-py/blob/main/siwe/siwe.py#L283 - which has the same options options, but aren't currently used.

Reply:
Thanks for the context, I was just making a general observation. But I think I'm convinced now that we don't need to do any other steps for authentication on the backend side. It should be enough that we require users to provide a way to validate `domain` and `resources` using the contract conditions.

I think this should hold as long our adopters don't try doing anything funny like reusing existing DIDs issues for something else and not for TACo specifically.

Reply:
> I think this should hold as long our adopters don't try doing anything funny like reusing existing DIDs issues for something else and not for TACo specifically.

Yeah, in this case do we/should we force the "domain" (or some other individual property) value in the SIWE message to be TACo-specific... that we can check by default.

Reply:
Another option, if we want to allow SIWE messages originally signed for other uses that are being reused (not sure if we want to do that but providing options ?) while having some guardrails, is that we could consistently enforce a freshness check relative to the `issued-at` value in the SIWE message - <https://docs.login.xyz/general-information/siwe-overview/eip-4361#message-field-descriptions>.

So, if `current time` > the `issued-at` time + `<some default constant offset>` (eg. 2 hours), then the authentication of the provided SIWE message fails.

Reply:
See [6421aed](https://github.com/nucypher/nucypher/pull/3502/commits/6421aedb8c2e42f3e60a2382321cfda5b29bf78d)

Reply:
> Yeah, in this case do we/should we force the "domain" (or some other individual property) value in the SIWE message to be TACo-specific (?)... that we can check by default. Basically SIWE messages must be TACo-specific.

I think it makes a lot of sense. Because I imagine SIWE users already doing some checks and balances to ensure that the DID relates to their application specifically. But issuing auth for said app is not the same as issuing auth for TACo within that dapp, and I think that could leak to some issues.

Reply:
> Anyone can sign a message with a TACo-specific property in it (eg. not using taco-web)

What exactly can 'anyone' do here, with regard to accessing data? If they also fulfill the conditions specified, are they spoofing anything?

Reply:
> What exactly can 'anyone' do here, with regard to accessing data? If they also fulfill the conditions specified, are they spoofing anything?

If I'm understanding correctly, "signing a message with a TACo-specific property" is not bad per se. It's more like a feature that we don't know if we want for TACo.

The use case is more like a user signing a message with some other application, and reusing this signature for TACo. There is no spoofing, simply they are using the same signature for two different products.

Reply:
@arjunhassard , 

what @manumonti said above is spot on. There's no spoofing of accounts to gain access, it's more about whether we need a specific SIWE message for TACo or not.

The question is, is TACo its own application OR a service used by an application as it pertains to SIWE? SIWE messages are seemingly for application-level interactions, but TACo doesn't work that way - it has a client-side and a server-side that are "disconnected" (different memory and execution space).

If TACo is its own application, sure, we can add a taco-specific value to the message, but SIWE is just a message template. Because the client-side and server-side are disconnected, you can't guarantee that the message came from using the `taco-web`/`taco-auth` library - any software can generate the same template message, so what are we gaining here?

In the latter case (TACo is just a service and not an app itself) from a UX perspective, signing in to an application using SIWE and then having to sign again using SIWE for TACo decryption could be annoying, in that the user has to sign a different message in addition to to what they already signed for the application? As a service used by the app, decryption using TACo (base functionality for `:userAddress`) just cares about the user is who they say they are with a reasonable message (not stale/expired). That could be the same application sign-in message. Any add-on conditions on the encrypted data may care about more than that, e.g. that the resource being accessed is allowed based on the "resources" value in SIWE message etc., but base TACo doesn't actually care or have the awareness to care. That awareness seems more like an additional condition on the data.

With all of that said, I'm not sure about the best practice for SIWE on the client side. But that is something to think about.

(cc @cygnusv )